cli/azd/docs/adding-command.md

```diff
@@ -0,0 +1,211 @@
+# Adding an azd command
+
+Read this if you are planning to add or update a console command to azd. 
+
+## Design
+
+Azd uses the [go cobra library](https://github.com/spf13/cobra) to describe the commands it support. However, with the introduction of azd-hooks, azd required to create links and properties which are beyond cobra's command regular expectations.
+
+The [cmd](https://github.com/Azure/azure-dev/blob/main/cli/azd/cmd/root.go#L4) package in `azd` uses a higher order component called [ActionDescriptor](https://github.com/Azure/azure-dev/blob/main/cli/azd/cmd/actions/action_descriptor.go). This component defines a command beyond the cobra's command scope. It describes options, flags, middleware, and relations between other actions. After creating a `tree` of actions, the root node can be used to generate a cobra command `tree` from it.
+
+## Adding top level command
+
+The `top level action descriptor` from azd is called `root`. It is created as part of the `NewRootCmd()` implementation. It is the starting phase of creating the cobra command hierarchy. In order to add a new top level command, you need to `add` a child *action descriptor* to `root` like:
+
+```golang
+root.Add("command-name", &actions.ActionDescriptorOptions{
+    Command:          *cobra.Command,
+    ActionResolver:   any,
+    FlagsResolver:    any,
+    DisableTelemetry: bool,
+    OutputFormats:    []output.Format,
+    DefaultFormat:    output.Format,
+    HelpOptions:      ActionHelpOptions,
+    GroupingOptions:  CommandGroupOptions
+})
```

Comment:
Thank you so much for putting into this into a doc. As I look at this, I can't help but think that we've gone off-path in complexity here.

I'm just putting this down as a mental note and a topic worthy of discussion.

In my mind, building a command requires 3 things:
1. An `action` to run
2. A `command` that contains the end-user design of the command. This has help, grouping, formats, flags.
3. `Options` that control preRunE or postRunE logic

Reply:
> I can't help but think that we've gone off-path in complexity here.

Creating the `ActionResolver` and the `FlagsResolver` is crazy!!  hahaha,,  the signature is `func() any`.... so there are no hints about they need to return `Actions.Action`.  I don't even know what happen if we add an actionResolver which returns something else, like `string`...  ( I am guessing it would fail on run-time while trying to construct the command... )

> A command that contains the end-user design of the command. This has help, grouping, formats, flags.

Yeah, but `cobra.command` is not enough for this... I think that' why we ended up introducing the action descriptor...

Reply:
@vhvb1989 Do you think this would been simpler if we just removed action, desciptors and went back to just `cobra.Command`?

I was thinking if we could just surround `cobra.Command` with a command `struct`